diff --git a/common-src/Makefile.am b/common-src/Makefile.am
index e51fde5..b62fb31 100644
--- a/common-src/Makefile.am
+++ b/common-src/Makefile.am
@@ -142,6 +142,7 @@ genversion_LDADD = $(libamanda_la_LIBADD)	\
     	versuff.lo				\
 	../gnulib/libgnu.la
 
+genversion.@OBJEXT@: $(genversion_SOURCES) genversion.h
 genversion.h: $(top_builddir)/config.status
 	-rm -f $@ $@.new
 	echo '#define CC "$(CC)"' > $@.new
